src/jsifier.js

```diff
@@ -45,7 +47,8 @@ function JSify(data, functionsOnly) {
     // else. This lets us not hold any strings in memory, we simply print
     // things out as they are ready.
 
-    var shellParts = read(shellFile).split('{{BODY}}');
+    var thirdParty = thirdPartyFiles.map(function(f) { return read(f) }).join('\n');
+    var shellParts = read(shellFile).replace('{{THIRD_PARTY}}', thirdParty).split('{{BODY}}');
```

Comment:
Got it, makes sense. Well, as an alternative, if we wanted to just steal the specific functions instead of using a dependency, below is what that would look like.

The original version of libsodium.js's `to_string` function was written by me and would've been similar to the above, but based on https://github.com/jedisct1/libsodium.js/issues/57 it seems that some sort of issue with handling large byte arrays was discovered and fixed, which would probably also become a problem for some emscripten users if we were to start from what I have above.

I don't mind working on making those implementations more robust and/or searching around for well tested public domain implementations if you'd prefer not to do this, but so we have the option: @jedisct1, would it be possible for the following functions from libsodium.js to be gifted to emscripten?

```
function to_string(bytes) {
	if (typeof TextDecoder === "function") {
		return new TextDecoder("utf-8", {fatal: true}).decode(bytes);
	}

	var toStringChunkSize = 8192,
		numChunks = Math.ceil(bytes.length / toStringChunkSize);
	if (numChunks <= 1) {
		try {
			return decodeURIComponent(escape(String.fromCharCode.apply(null, bytes)));
		}
		catch (_) {
			throw new TypeError("The encoded data was not valid.");
		}
	}
	var totalString = '';
	var sequenceReadOffset = 0;
	for (var i = 0; i < numChunks; i++) {
		var currentChunk =
			Array.prototype.slice.call(bytes,
										i * toStringChunkSize + sequenceReadOffset,
										(i + 1) * toStringChunkSize + sequenceReadOffset);
		//Depending on how much we have shifted
		if (currentChunk.length == 0) {
			continue;
		}

		//Checking that we didn't cut the buffer in the middle of a UTF8 sequence.
		//If we did, remove the bytes of the "cut" sequence and
		//decrement sequenceReadOffset for each removed byte
		var sequenceDetectionComplete,
			sequenceIndex = currentChunk.length,
			sequenceLength = 0;

		//This loop will read the chunk from its end, looking for sequence start bytes
		do {
			sequenceIndex--;
			var currentByte = currentChunk[sequenceIndex];

			if (currentByte >= 240) { //Beginning of a 4-byte UTF-8 sequence
				sequenceLength = 4;
				sequenceDetectionComplete = true;
			} else if (currentByte >= 224) { //Beginning of a 3-byte UTF-8 sequence
				sequenceLength = 3;
				sequenceDetectionComplete = true;
			} else if (currentByte >= 192) { //Beginning of a 2-byte UTF-8 sequence
				sequenceLength = 2;
				sequenceDetectionComplete = true;
			} else if (currentByte < 128) { //A one byte UTF-8 char
				sequenceLength = 1;
				sequenceDetectionComplete = true;
			}
			//The values between [128, 192[ are part of a UTF-8 sequence.
			//The loop will not exit in that case, and will iterate one byte backwards instead
		} while (!sequenceDetectionComplete);

		var extraBytes = sequenceLength - (currentChunk.length - sequenceIndex);
		for (var j = 0; j < extraBytes; j++) {
			sequenceReadOffset--;
			currentChunk.pop();
		}

		totalString += to_string(currentChunk);
	}
	return totalString;
}

function from_base64(sBase64, nBlocksSize) {
	function _b64ToUint6(nChr) {
		return nChr > 64 && nChr < 91 ?
			nChr - 65 : nChr > 96 && nChr < 123 ?
			nChr - 71 : nChr > 47 && nChr < 58 ?
			nChr + 4 : nChr === 43 ?
			62 : nChr === 47 ?
			63 :
			0;
	}

	var sB64Enc = sBase64.replace(/[^A-Za-z0-9\+\/]/g, ""),
		nInLen = sB64Enc.length,
		nOutLen = nBlocksSize ? Math.ceil((nInLen * 3 + 1 >> 2) / nBlocksSize) * nBlocksSize : nInLen * 3 + 1 >> 2,
		taBytes = new Uint8Array(nOutLen);

	for (var nMod3, nMod4, nUint24 = 0, nOutIdx = 0, nInIdx = 0; nInIdx < nInLen; nInIdx++) {
		nMod4 = nInIdx & 3;
		nUint24 |= _b64ToUint6(sB64Enc.charCodeAt(nInIdx)) << 18 - 6 * nMod4;
		if (nMod4 === 3 || nInLen - nInIdx === 1) {
			for (nMod3 = 0; nMod3 < 3 && nOutIdx < nOutLen; nMod3++, nOutIdx++) {
				taBytes[nOutIdx] = nUint24 >>> (16 >>> nMod3 & 24) & 255;
			}
			nUint24 = 0;
		}
	}
	return taBytes;
}
```